modules/packages/Image.chpl

```diff
@@ -357,4 +370,299 @@ module Image {
+        import super.STBImageHelper;
+        STBImageHelper.writePng(outfile, pixels);
+      } else if impl == Implementation.libpng {
+        compilerError("libpng is not yet supported");
+      } else {
+        compilerError("Unknown PNG implementation");
+      }
+    }
+    proc read(infile) throws {
+      if impl == Implementation.stb_image {
+        import super.STBImageHelper;
+        return STBImageHelper.readPng(infile);
+      } else if impl == Implementation.libpng {
+        compilerError("libpng is not yet supported");
+      } else {
+        compilerError("Unknown PNG implementation");
+      }
+    }
+  }
+
+  private module JPGHelper {
+    enum Implementation {
+      stb_image
+    }
+    config param impl = Implementation.stb_image;
+    proc write(outfile, pixels: [?d]) throws {
+      if impl == Implementation.stb_image {
+        import super.STBImageHelper;
+        STBImageHelper.writeJpg(outfile, pixels);
+      } else {
+        compilerError("Unknown JPG implementation");
+      }
+    }
+    proc read(infile) throws {
+      if impl == Implementation.stb_image {
+        import super.STBImageHelper;
+        return STBImageHelper.readJpg(infile);
+      } else {
+        compilerError("Unknown JPG implementation");
+      }
+    }
+  }
+
+  private module STBImageHelper {
+    require "ImageHelper/stb_image_helper.h";
+
+    private use CTypes;
+    private use IO;
+    private use Image;
+
+    /*private*/ type stbi_write_func = c_fn_ptr;
+    private extern proc stbi_write_png_to_func(func: stbi_write_func,
+                               context: c_ptr(void),
+                               w: c_int,
+                               h: c_int,
+                               comp: c_int,
+                               data: c_ptrConst(void),
+                               stride_in_bytes: c_int): c_int;
+
+    private extern proc stbi_write_jpg_to_func(func: stbi_write_func,
+                               context: c_ptr(void),
+                               w: c_int,
+                               h: c_int,
+                               comp: c_int,
+                               data: c_ptrConst(void),
+                               quality: c_int): c_int;
+
+    private proc stbi_writeFuncLocking(context: c_ptr(void),
+                             data: c_ptrConst(void),
+                             size: c_int) {
+      const outfile = (context:c_ptr(fileWriter(locking=true))).deref();
+      try! outfile.writeBinary(data:c_ptrConst(uint(8)):c_ptr(uint(8)), size);
+    }
+    private proc stbi_writeFunc(context: c_ptr(void),
+                             data: c_ptrConst(void),
+                             size: c_int) {
+      const outfile = context:c_ptr(fileWriter(locking=false));
+      try! outfile.writeBinary(data:c_ptrConst(uint(8)), size);
+    }
+
+    // rewrite pixels into the right format for stb_image
+    private proc writeCommon(outfile, pixels: [?dom]) throws {
```

Comment:
with `outfile`'s type unspecified, the `c_ptrToConst(outfile)` below is relatively confusing to read. Also, the lack of a return type annotation on this 5-tuple-returning function doesn't help reading it.

Reply:
If you're using a void pointer as "data" for callbacks in STB -- looks like you are -- you almost certainly want to use an explicit `ref` intent on the argument lest you get lifetime wonkiness. E.g., if you context is an int (hypothetically), it'll be copied in, and a pointer to it will not be valid after this function returns.

It looks like `context` at this time is only ever expected to be a `fileWriter` (good reason to explicitly annotate the type). Today, the intent for it will be `ref`, so you will not run into trouble with lifetimes. However, the spec i believe specifies the intent as `const`, which may or may not perform a copy. If it does, you'd be creating a pointer to a local copy of the record.

Reply:
I made this more consistent for both reading and writing